petsc: refactor and add petsc4py support

qbisi opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
1. refactor configure phase.
2. add some optional dependency which will be used by firedrake/fenics.
3. petsc4py enabled (petsc 3.21.4->3.22.3 needed for python313Packages.petsc4py module support)

petsc install petsc4py under $out/lib which is not the desired behavior $out/lib/${python3.sitePackages} and cannot be
configured via flags or envrionment.
So i patched the configure file in petsc source code.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

qbisi committed:
petsc: simplify configurePhase

qbisi committed:
petsc: reformat with nixfmt-rfc-style

qbisi committed:
python312Packages.petsc4py: init at 3.22.3

qbisi committed:
petsc: add standalone metis support

qbisi committed:
petsc: add ptscotch support

qbisi committed:
petsc: add scalapack and mumps support

qbisi committed:
petsc: add withFullDeps option

qbisi committed:
petsc: rename some optional dependency flags

NickCao commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -67,29 +69,22 @@ stdenv.mkDerivation rec {
+  ] ++ lib.optionals petsc-withp4est [
+    "--with-p4est=1"
+    "--with-zlib-include=${zlib.dev}/include"
+    "--with-zlib-lib=[-L${zlib}/lib,-lz]"
```
First time seeing this syntax, would you mind sharing a link to relevant documentations?

qbisi replied:
This syntax is only used by petsc configure script to bypass space in configure flags.

qbisi replied:
Yes we can. But for petsc, the bracket syntax have avoided the use of space, So there is no need to set __structuredAttrs=true here.

NickCao commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -69,6 +78,10 @@ stdenv.mkDerivation rec {
     ++ lib.optional withScalapack scalapack
     ++ lib.optional withMumps mumps_par;
 
+  propagatedBuildInputs = lib.optional withPetsc4py python3.pkgs.numpy;
```
Preferably use `python3Packages` over `python3.pkgs`.

NickCao commented on pkgs/by-name/pe/petsc/fix-petsc4py-install-prefix.patch:
```diff
@@ -0,0 +1,23 @@
+     import os
+-    installLibPath = os.path.join(self.installDir, 'lib')
++    from sys import version_info
++    installLibPath = os.path.join(self.installDir, 'lib', f"python{version_info.major}.{version_info.minor}" ,'site-packages')
```
Or use `python3.sitePackages` and `replaceVars`?

qbisi replied:
This is only a nixpkgs related patch. I will try the more nix way to replace the installdir.

NickCao commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -29,11 +29,11 @@ assert petsc-withp4est -> p4est.mpiSupport;
-  version = "3.21.4";
+  version = "3.22.3";
 
   src = fetchzip {
```
Can we fetch the source code from https://gitlab.com/petsc/petsc, which would make future reviews easier.

qbisi replied:
This tarball link is the alternative download method  recommended by petsc, see https://petsc.org/release/install/download. An other method is to fetch the git release in gitlab. Both are welcomed.

NickCao replied:
If that's recommended by upstream, let's keep it as is.

sheepforce reviewed (approved):
nixpkgs-review says PFLOTRAN is the only package broken by this update, which is an acceptable breakage for me. I haven't tested the python bindings, but the rest looks good to me.

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 379406`

---
### `x86_64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>pflotran</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 9 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
  </ul>
</details>

qbisi commented:
> nixpkgs-review says PFLOTRAN is the only package broken by this update, which is an acceptable breakage for me. I haven't tested the python bindings, but the rest looks good to me.
> 
> ## `nixpkgs-review` result
> Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).
> 
> Command: `nixpkgs-review pr 379406`
> 
> ### `x86_64-linux`
> ❌ 1 package failed to build:
> ✅ 9 packages built:

Thank you for your tests, maybe we can apply the patch to pflotran which fix the build or wait for next release.
I am also wondering how many packages depending on petsc want the optimized feature. 
Maybe we can set petsc-optimized=true by default.
Petsc has many optional dependency, some are commonly used like hdf5, would you like some petsc varient (petsc_minimal petsc petsc_full) just like ffmpeg does.

markuskowa commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -4,92 +4,153 @@
-    )
-  '' + lib.optionalString hdf5-support ''
-    configureFlagsArray+=(
+      "--with-zlib-lib=[-L${zlib}/lib,-lz]"
```
```suggestion
      "--with-zlib-lib=[-L${lib.getLib zlib}/lib,-lz]"
```

markuskowa commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -4,92 +4,153 @@
-      "--with-hdf5-lib=-L${hdf5}/lib -lhdf5"
-    )
-  '';
+      "--with-hdf5-lib=[-L${hdf5}/lib,-lhdf5]"
```
```suggestion
      "--with-hdf5-lib=[-L${lib.getLib hdf5}/lib,-lhdf5]"
```

markuskowa commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -4,92 +4,153 @@
+    ++ lib.optionals withPtscotch [
+      "--with-ptscotch=1"
+      "--with-ptscotch-include=${scotch.dev}/include"
+      "--with-ptscotch-lib=[-L${scotch}/lib,-lptscotch,-lptesmumps,-lptscotchparmetisv3,-lptscotcherr,-lesmumps,-lscotch,-lscotcherr]"
```
```suggestion
      "--with-ptscotch-lib=[-L${lib.getLib scotch}/lib,-lptscotch,-lptesmumps,-lptscotchparmetisv3,-lptscotcherr,-lesmumps,-lscotch,-lscotcherr]"
```

markuskowa commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -4,92 +4,153 @@
   mpiCheckPhaseHook,
-  petsc-withp4est ? false,
-  hdf5-support ? false,
+  withP4est ? withFullDeps,
```
This is a breaking change. 
CC @cburstedde

qbisi replied:
petsc-withp4est is only found/used in petsc,  hdf5-support is only found/used in petsc and pflotran , i have split this rename change to a indiviual commit, if @cburstedde @sheepforce are happy with this change, we can adopt this commit.

sheepforce replied:
Is the critique here that those were disabled by default and now enabled by default? For me that is perfectly fine, however, I'm using it only in one package and enable HDF5 there, so my opinion on that may not be the most relevant one :smile:

qbisi replied:
They are not enabled  by default in petsc since withFullDeps=false by default, however withFullDeps=true is set in petsc4py to support fenics/firedrake.  If both petsc_minimal, petsc, and petsc_full are provided what would you like to choose in your projects.

qbisi committed:
petsc: use lib.{getDev,getLib} instead *.{dev,lib}

sheepforce commented:
> Thank you for your tests, maybe we can apply the patch to pflotran which fix the build or wait for next release.

The patch could proof difficult. I'm not familiar with the internals of PFLOTRAN and the newer versions of Petsc lead to rank mismatches in many arrays. I think it is perfectly fine to break PFLOTRAN for the moment and fix it at the next release. It's not clear yet when the next release will happen and I don't want to hold back the petsc updates.

qbisi commented:
> > Thank you for your tests, maybe we can apply the patch to pflotran which fix the build or wait for next release.
> 
> The patch could proof difficult. I'm not familiar with the internals of PFLOTRAN and the newer versions of Petsc lead to rank mismatches in many arrays. I think it is perfectly fine to break PFLOTRAN for the moment and fix it at the next release. It's not clear yet when the next release will happen and I don't want to hold back the petsc updates.

python313package for petsc4py is not mendatory for now（I used to think it is required to init petsc4py ),also lateset version of petsc does not appeal to me , maybe we can hold up the version till the fixup of pflotran or next major release of petsc.

markuskowa commented:
Can you please squash the review commits into their respective commits.

qbisi commented:
You mean squash the last three review commit from your suggestion ? @markuskowa

I am going to hold on the petsc version for this pr to be compatiable with pflortran (so only python312package.petsc4py is provided). @sheepforce

markuskowa commented:
> You mean squash the last three review commit from your suggestion ? @markuskowa

Yes.

qbisi committed:
petsc: reorder build options and external dependency

qbisi committed:
petsc: add hdf5 fortran binding

qbisi committed:
petsc: set optimized and withHdf5 default to true

Petsc is commonly optimized built with hdf5 support to be used
as a component of other software and project.
See
    1. https://sources.debian.org/src/petsc/3.22.2%2Bdfsg1-1/debian/rules/#L154
    2. https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=petsc

qbisi commented:
Squashing the last three commit into the first one cause a great rebase conflict due to the nix-format commit in the way.
So i just squashing them into a standalone one.

Note for the last two commit 
- hdf5 -> hdf5-fortran-mpi : petsc offers fortran and mpi binding thus the latter is more appropriate
- hdf5 -> by defautl true: commonly used, one can see they are compiled in by default in Debian Source and Aur
      1. https://sources.debian.org/src/petsc/3.22.2%2Bdfsg1-1/debian/rules/#L154
      2. https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=petsc
- optimized -> by default true: Software and Projects who uses petsc would always like a release build

markuskowa commented:
@sheepforce okay with you now?

sheepforce reviewed (approved):
> @sheepforce okay with you now?

Sure, LGTM! :sheep:
